syntax = "proto3";

package NCloud.NBlockStore.NProto;

option java_package = "ru.yandex.cloud.blockstore.proto";
option go_package = "a.yandex-team.ru/cloud/blockstore/public/api/protos";

////////////////////////////////////////////////////////////////////////////////
// Encryption mode

enum EEncryptionMode
{
    NO_ENCRYPTION = 0;
    ENCRYPTION_AES_XTS = 1;
    ENCRYPTION_TEST = 2;
}

////////////////////////////////////////////////////////////////////////////////
// KMS spec

message TKmsSpec
{
    // Id of key encryption key.
    string KekId = 1;

    // Encrypted data encryption key.
    bytes EncryptedDEK = 2;
}

////////////////////////////////////////////////////////////////////////////////
// Encryption key path

message TKeyPath
{
    oneof Path
    {
        // Spec to getting the key from KMS.
        TKmsSpec KmsSpec = 1;

        // Identifier of the keyring containing the key.
        uint32 KeyringId = 2;

        // Path to the file containing the key.
        string FilePath = 3;

        // Data encryption key.
        bytes DEK = 4;
    }
}

////////////////////////////////////////////////////////////////////////////////
// Encryption key

message TEncryptionKey
{
    // Encryption mode.
    EEncryptionMode Mode = 1;

    // Encryption key path.
    TKeyPath KeyPath = 2;
}

////////////////////////////////////////////////////////////////////////////////
// Encryption spec

message TEncryptionSpec
{
    oneof KeyParam
    {
        // Encryption key.
        TEncryptionKey Key = 1;

        // Encryption key hash.
        bytes KeyHash = 2;
    }

    // Disable encryption for snapshot.
    bool DisableEncryption = 3;
}
